core/sail/shacl/src/main/java/org/eclipse/rdf4j/sail/shacl/VerySimpleRdfsBackwardsChainingConnection.java

```diff
@@ -47,17 +50,34 @@
 	public boolean hasStatement(Resource subj, IRI pred, Value obj, boolean includeInferred, Resource... contexts)
 			throws SailException {
 
-		boolean hasStatement = super.hasStatement(subj, pred, obj, includeInferred, contexts);
+		boolean hasStatement = super.hasStatement(subj, pred, obj, false, contexts);
+
+		if (hasStatement) {
+			return true;
+		}
 
 		if (rdfsSubClassOfReasoner != null && includeInferred && obj != null && obj.isResource()
 				&& RDF.TYPE.equals(pred)) {
-			return hasStatement | rdfsSubClassOfReasoner.backwardsChain((Resource) obj)
-					.stream()
-					.map(type -> super.hasStatement(subj, pred, type, false, contexts))
-					.reduce((a, b) -> a || b)
-					.orElse(false);
+
+			Set<Resource> types = rdfsSubClassOfReasoner.backwardsChain((Resource) obj);
+			if (types.size() > 10) {
+				try (CloseableIteration<? extends Statement, SailException> statements = super.getStatements(subj,
+						RDF.TYPE, null, false, contexts)) {
+					return statements.stream()
```

Comment:
So if I understand correctly, this is based on the assumption that typically, the number of (explicit) type statements for a given subject is going to be fewer than 10, correct? 

That's a sensible enough rule of thumb, though of course it may fluctuate wildly between domains. As a quick performance fix this is ok with me, but I'm thinking for a future larger improvement: rather than putting a hardcoded limit in, wouldn't it make sense to tie this in with the store statistics? I know those statistics are currently unaccessible outside the store implementation itself, might a thing to fix in 4.0.

Reply:
This is bang on!

I thought 10 was a decent size, but I'm sure this isn't the case for all users. 

Access to statistics would be great! We could always generate out of statistics here, but I feel  it would just add unnecessary complexity.

Reply:
Opened an issue: https://github.com/eclipse/rdf4j/issues/3249